core/arch/arm/sm/sm_a32.S

```diff
@@ -166,6 +166,10 @@ UNWIND(	.cantunwind)
 	b	.sm_exit
 
 .smc_from_nsec:
+	/* Invalidate the branch predictor. */
+	write_bpiallis
```

Comment:
You're right, I'll fix.

Reply:
This is late to apply the fix as the BTB speculation can occur on the unconditional branch at Line 258. BPIALL needs to be done prior to any branch instruction in Secure world. Its not easy to do this but we have an upcoming patch in ARM TF which will demonstrate this.

Reply:
@soby-mathew , so you are saying this should be done before actually doing any jump from the vector table. Will this also affect the [thread vector table](https://github.com/OP-TEE/optee_os/blob/486754e82582c0617518fddace459b5fdb5ca8d9/core/arch/arm/kernel/thread_a64.S#L154-L167)? I'd guess that it doesn't affect that, agree?

Reply:
Interesting, does it involve multiple vectors?

Reply:
Yes, the vector table handling the switch from normal to secure must do the BPIALL prior to any branch instruction. In Arm32 when OPTEE is the Monitor code, it needs to do this.

The thread vector need not do this because ARM TF in EL3 would have already done the fix prior to reaching the thread vector.

Reply:
@soby-mathew speculation can occur on an *unconditional* branch??? Sorry for insisting, but this seems counter-intuitive to me. I mean, for such a branch there is no decision to be made whether or not the branch will be taken... so why would the BTB be used at all? Ah, maybe because the actual address of the branch might not be readily available, so the CPU would speculate while waiting for the address to be resolved?

Reply:
@jforissier, this was counter intuitive to me as well. I am not sure of the details, but the CPU can speculate even for an unconditional branch.

Reply:
@soby-mathew OK, thanks.

Reply:
@jenswi-linaro, the way we do it is to have ORR (Or ADD) instruction at each vector entry to encode a bit pattern into an available General Purpose register. We use the `sp` for this and ensure the stack/whatever sp is pointing to is aligned so that the LSB can be used for the bit pattern encoding. Then the BPIALL is done and the bit pattern is extracted to get the vector entry. Finally we jump to the correct handler corresponding to the vector entry.

If the `sp` is pointing to address that is naturally aligned to 256 bytes then ORR can be used to encode the bit pattern (one bit per vector entry). Else ADD can be used to encode values from 0 to 7 for each vector entry and the `sp` needs to point to an address that is 8 byte aligned.

That's the general idea.

Reply:
Aha, clever. Thanks, I'll do that.

Reply:
All branches are subject to prediction on some Arm CPUs.

ARM Technical Reference Manual for A57 on predicted instructions:

>**6.5.1 Predicted and non-predicted instructions**
This section describes the instructions that the processor predicts.
>
>Unless otherwise specified, the list applies to A32, T32, and A64 instructions. As a general rule, the branch prediction hardware predicts all branch instructions regardless of the addressing mode, including:
>* Conditional branches.
>* Unconditional branches.
>* Indirect branches.
>* Branches that switch between ARM and Thumb states.
>* PC destination data processing operations.
>* BXJ, because of the inclusion of the trivial Jazelle implementation, this degenerates to a BX instruction. There is no BXJ instruction in A64.
>
>However, the following branch instructions are not predicted:
>* AArch32 instructions with the S suffix are not predicted because they are typically used when returning from exceptions and have side effects that can change privilege mode and Security state.
>* All mode or Exception level changing instructions.
>
>In Thumb state, you can make a branch that is normally encoded as unconditional conditional by including an If-Then (IT) block. It is then treated as a normal conditional branch.

see http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0488h/way1382448710808.html

Reply:
Jens - we're also testing a version that has less constraint on the SP alignment and only needs 3 bits of SP to encode the vector before invalidating the branch predictor.

Reply:
@athoelke thanks for the reference, quite helpful.

I suppose this section will need some rework however:
 http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0488h/way1382448710808.html
;-)